docs/design/2025-03-17-sys-session-manage-enhancement.md

```diff
@@ -0,0 +1,167 @@
+
+Some PR such as [#59546](https://github.com/pingcap/tidb/pull/59546) attempt to address these issues by introducing a `Destroy` method for the session pool. However, this approach does not fully resolve the problems and still makes it difficult to ensure that developers call the correct method.
+
+This document proposes a more effective solution for managing internal system sessions to achieve the following goals:
+
+- Prevent concurrent usage across modules: Once a session is returned to the pool, it should no longer be used to avoid potential conflicts from concurrent access.
+- Simplify session lifecycle management: Ensure that sessions are always either returned to the pool or properly closed after use, eliminating the risk of memory leaks.
+- Enforce session state reset: When a session is returned to the pool, its internal state should be forcibly reset to prevent unexpected behaviors in subsequent usage.
+
+## Design
+
+### Pool
+
+We need to introduce a new enhanced session pool to manage internal system sessions. The new session pool will provide the following methods:
+
+```go
+type Pool struct {
+    ...
+}
+
+func (p *Pool) get() (*Session, error) {
+    ...
+}
+
+func (p *Pool) put(s *Session) {
+    ...
+}
+
+func (p *Pool) WithSession(func(*Session) error) error {
+    ...
+}
+```
+
+The most important method of `Pool` is `WithSession` which is introduced to replace the `Get` and `Put` methods of the current session pool. The `WithSession` method accepts a function as an argument, which will be executed with a session retrieved from the pool. After the function completes, the session will be automatically returned to the pool. This design ensures that the session is always returned to the pool after use, preventing memory leaks.
+
+Let's see how `WithSession` works:
+
+```go
+func (p *Pool) WithSession(func(*Session) error) (err error) {
+    se, err := p.get()
+    if err != nil {
+        return err
+    }
+    
+    defer func() {
+        if r := recover(); r != nil {
+            // Always destroy the session when panic to avoid undetermined state.
+            se.Destroy()    
+            panic(r)
+        }
+        p.put(se)
+    }()
+    return f(se)
+}
+```
+
+We can see that the `WithSession`n method always calls `p.put(se)` within a defer function to ensure the session is returned to the pool. One exception to this occurs when a panic happens. In such cases, the session is destroyed instead of being returned to the pool, as it may be in an undefined state, making it safer to discard rather than reuse.
```

Comment:
```suggestion
We can see that the `WithSession` method always calls `p.put(se)` within a defer function to ensure the session is returned to the pool. One exception to this occurs when a panic happens. In such cases, the session is destroyed instead of being returned to the pool, as it may be in an undefined state, making it safer to discard rather than reuse.
```